### DISCUSSION TITLE
Custom LLM integration with Voice Agent

### DISCUSSION THREAD

**Discussion by satyam-veris** [deepgram/community#1034] [General help] [answered]
I'm trying to integrate my custom LLM with Deepgram Voice Agent. When using their default providers (OpenAI/Anthropic), everything works fine. With OpenAI, I see this response format in the WebSocket messages:
```json
{
    "type": "ConversationText",
    "role": "assistant", 
    "content": "How can I help you?"
}
```

I've set up my custom LLM REST endpoint at /chat in my localhost and exposed it publicly using NGROK.  and configured the agent:

```json
{
  "think": {
    "provider": {
      "type": "custom",
      "url": "my-server/v1/chat/completions"
    }
  }
}
```

My LLM receives and processes requests successfully (confirmed via logs), but the responses never appear in the WebSocket messages. What's the correct response format for custom LLM integration?

My server.js file looks like

```javascript
app.post('/chat', async (req, res) => {
  try {
    const { messages, thread_id = 'default' } = req.body;
    const lastMessage = messages[messages.length - 1];
    const response = await callAgent(mongoClient, lastMessage.content, thread_id);
    console.log('response', response);
    // Format response for Deepgram
    const finalResponse = {
      id: `chatcmpl-${Date.now()}`,
      object: 'chat.completion',
      created: Math.floor(Date.now() / 1000),
      model: "custom-langgraph-agent",
      usage: {
        prompt_tokens: 0,
        completion_tokens: 0,
        total_tokens: 0
      },
      choices: [{
        message: {
          role: 'assistant',
          content: response,
          type: 'ConversationText',
        },
        finish_reason: "stop",
        index: 0
      }]
    };

    res.json(finalResponse);

  } catch (error) {
    console.error('Error:', error);
    res.status(500).json({ error: { message: 'Internal server error' } });
  }
});
```

Sample config for deepgram Agent:

```javascript

const baseConfig = {
  type: "SettingsConfiguration",
  audio: audioConfig,
  agent: {
    listen: { model: "nova-2" },
    speak: { model: "aura-asteria-en" },
    think: {
      provider: {
        type: "custom",
        url: '<server_url>/chat',
        headers: [
          { key: 'Content-Type', value: 'application/json' }
        ],
      },
      model: "custom-langgraph-agent",
      // provider: { type: 'open_ai' },
      // model: 'gpt-4o-mini'
    },
  },
};
```

I am using this repository as a interface https://github.com/deepgram-devs/deepgram-voice-agent-demo.

Environment:
- Custom LLM running on Express server
- Using ngrok for testing
- Deepgram Voice Agent in browser

**Comment by deepgram-community** [comment]
Thanks for asking your question. Please be sure to reply with as much detail as possible so the community can assist you efficiently.
<sub>Consider joining our [Discord community](https://discord.gg/deepgram) for more opportunity to engage with your fellow Deepgram users. You can earn points which can be redeemed for cool stuff by being active in our communities!</sub>

**Comment by deepgram-community** [comment]
Hey there! It looks like you haven't connected your GitHub account to your Deepgram account. You can do this at https://community.deepgram.com - being verified through this process will allow our team to help you in a much more streamlined fashion.

**Comment by deepgram-community** [comment]
It looks like we're missing some important information to help debug your issue. Would you mind providing us with the following details in a reply?

- A request ID that triggered your error or issue.

**Reply by alexsam9866** [reply]
Deepgram won’t handle OpenAI-style nested JSON, it needs a simple, flat response like:

`{
  "type": "ConversationText",
  "role": "assistant",
  "content": "Your response here"
}`

If your custom LLM is sending something like choices[0].message, Deepgram will ignore it. You’ll need to adjust your endpoint to return just the inner message directly, no wrapping, no metadata. Strip it down to the flat object, and it’ll work as expected.

**Comment by deepgram-community** [comment, marked as answer]
You might want to check how you are setting your configuration for the custom model.

Our Docs give an example here: <https://developers.deepgram.com/docs/voice-agent-llm-models#passing-a-custom-llm-through-a-cloud-provider>

Also there is a working demo of doing this with Azure Open AI here which might help you: <https://github.com/deepgram-devs/voice-agent-azure-open-ai-services/blob/1ad0eee7d99cff521da0de9abaf247ad24a4e1a6/client.py#L41>


Though if you are running your LLM locally, that might make this a bit more challenging as I have not tried that yet.

>  I've set up my custom LLM REST endpoint at /chat in my localhost and exposed it publicly using NGROK.  and configured the agent:
There might be something on the NGROK side you need to look at to enable the connection.

> ![avatar](https://avatars.slack-edge.com/2023-02-20/4855338233600_fad9174023ee3f640ce1_32.png) This message was sent by John Vajda from Deepgram, via our community automation.

**Comment by itsjitender** [comment]
Hi
I don't want to use Azure Open AI & AWS Bedrock. I have my own Express API where it is interacting with OpenAI and get the response.
I can see request come to my custom API endpoint and I able to process same with Open AI but I still not figure out what response expect by deepgram to process that. It is not doing anything at all with open AI JSON output.
@satyam-veris  Did you able to run via your own Expess App? If yes, can you share the details.

deepgram- community : Do you have sample which work without Azure Open AI & AWS? Or you don't support any other Custom LLM beside these?

**Reply by satyam-veris** [reply]
Hi Jitendra,

I had tried multiple things but wasn't able to figure out how to use custom LLM with Deepgram. Then I thought to try out 11labs where they had a clear description about how to  structure your response for custom LLM usage. Eventually that worked with both 11labs and Deepgram.

![image](https://github.com/user-attachments/assets/4c99d0f7-2652-4458-a8d8-1b0025ed1d16)

Chat Completion API: https://platform.openai.com/docs/api-reference/chat/create

**Reply by itsjitender** [reply]
@satyam-veris  Thanks for quick response. I tried but still no luck yet.
My express sending below JSON response to deepgram. Can you please take a look

{
  "id": "chatcmpl-BNJkRiMOEGGxzs0nOZA4mIZQkUi1N",
  "object": "chat.completion",
  "created": 1744897155,
  "model": "gpt-4o-mini",
  "choices": [
    {
      "index": 0,
      "message": {
        "role": "assistant",
        "content": "It sounds like you're mentioning \"booking.\" Can you clarify what you're looking to book?",
        "refusal": null,
        "annotations": []
      },
      "logprobs": null,
      "finish_reason": "stop"
    }
  ],
  "usage": {
    "prompt_tokens": 194,
    "completion_tokens": 18,
    "total_tokens": 212,
    "prompt_tokens_details": {
      "cached_tokens": 0,
      "audio_tokens": 0
    },
    "completion_tokens_details": {
      "reasoning_tokens": 0,
      "audio_tokens": 0,
      "accepted_prediction_tokens": 0,
      "rejected_prediction_tokens": 0
    }
  },
  "service_tier": "default"
}

**Reply by satyam-veris** [reply]
You may take reference from this express server code which worked for me. I further fine-tuned it for my use case and customized the llm.

```js
require('dotenv').config();
const express = require('express');
const OpenAI = require('openai');
const app = express();

const OPENAI_API_KEY = process.env.OPENAI_API_KEY;
if (!OPENAI_API_KEY) {
    throw new Error("OPENAI_API_KEY not found in environment variables");
}

const oai_client = new OpenAI({ apiKey: OPENAI_API_KEY });

app.use(express.json());

app.post("/v1/chat/completions", async (req, res) => {
    const { messages, model, temperature = 0.7, max_tokens, stream = false, user_id } = req.body;
    console.log({ messages });
    const oai_request = {
        messages,
        model,
        temperature,
        stream,
        ...(max_tokens && { max_tokens }),
        ...(user_id && { user: user_id })
    };

    try {
        if (stream) {
            res.setHeader('Content-Type', 'text/event-stream');
            res.setHeader('Cache-Control', 'no-cache');
            res.setHeader('Connection', 'keep-alive');

            const stream = await oai_client.chat.completions.create(oai_request);
            console.log('stream response', stream);
            for await (const chunk of stream) {
                res.write(`data: ${JSON.stringify(chunk)}\n\n`);
            }
            res.write('data: [DONE]\n\n');
            res.end();
        } else {
            const completion = await oai_client.chat.completions.create(oai_request);
            console.log('completion response', completion);
            res.json(completion);
        }
    } catch (error) {
        console.error("An error occurred:", error);
        res.status(500).json({ error: 'Internal error occurred!' });
    }
});

const PORT = process.env.PORT || 8080;
app.listen(PORT, () => {
    console.log(`Server running on port ${PORT}`);
});
```

**Reply by itsjitender** [reply]
@satyam-veris Thank you very much.
The code provided by you finally worked for me.

**Reply by SoumyaNC** [reply]
Hi... I am facing similar issue. I am not using any LLM to get the output. I have my own logic to send the response back to deepgram. Request is coming to my endpoint but not sure in what format we should send back the response

**Reply by satyam-veris** [reply]
> Hi... I am facing similar issue. I am not using any LLM to get the output. I have my own logic to send the response back to deepgram. Request is coming to my endpoint but not sure in what format we should send back the response

You will need to stream your response to Deepgram. A normal JSON response won't work, Additionally you would also need to make sure the response structure matches the OpenAI Chat Completion API

**Reply by SoumyaNC** [reply]
> > Hi... I am facing similar issue. I am not using any LLM to get the output. I have my own logic to send the response back to deepgram. Request is coming to my endpoint but not sure in what format we should send back the response
> 
> You will need to stream your response to Deepgram. A normal JSON response won't work, Additionally you would also need to make sure the response structure matches the OpenAI Chat Completion API

Thanks for the response. But the thing is I am not using any external LLM like OPEN AI or anything for the output. I am generating the output based on the request content. I tried to match the response of OPEN AI Chat like this 

const response = {
    id: `chatcmpl-232`,
    object: 'chat.completion',
    created: Math.floor(Date.now() / 1000),
    model: model || 'gpt-4',
    choices: [
      {
        index: 0,
        message: {
          role: 'assistant',
          content: text
        },
        finish_reason: 'stop'
      }
    ],
    usage: {
      prompt_tokens: 9,
      completion_tokens: 6,
      total_tokens: 15
    },
    service_tier: "default"
  };

After that I am not able to understand how to stream the output back to Deepgram.

**Reply by satyam-veris** [reply]
> ```js
> if (stream) {
>             res.setHeader('Content-Type', 'text/event-stream');
>             res.setHeader('Cache-Control', 'no-cache');
>             res.setHeader('Connection', 'keep-alive');
> 
>             const stream = await oai_client.chat.completions.create(oai_request);
>             console.log('stream response', stream);
>             for await (const chunk of stream) {
>                 res.write(`data: ${JSON.stringify(chunk)}\n\n`);
>             }
>             res.write('data: [DONE]\n\n');
>             res.end();
>         }
> ```

You can take reference from the code I shared in the thread.

```js
if (stream) {
            res.setHeader('Content-Type', 'text/event-stream');
            res.setHeader('Cache-Control', 'no-cache');
            res.setHeader('Connection', 'keep-alive');

            const stream = await generateResponse(request); // should return response in OAI chat completion structure. return type: Array<response>
            console.log('stream response', stream);  
            for await (const chunk of stream) {
                res.write(`data: ${JSON.stringify(chunk)}\n\n`);
            }
            res.write('data: [DONE]\n\n');
            res.end();
}
```

**Reply by SoumyaNC** [reply]
> > ```js
> > if (stream) {
> >             res.setHeader('Content-Type', 'text/event-stream');
> >             res.setHeader('Cache-Control', 'no-cache');
> >             res.setHeader('Connection', 'keep-alive');
> > 
> >             const stream = await oai_client.chat.completions.create(oai_request);
> >             console.log('stream response', stream);
> >             for await (const chunk of stream) {
> >                 res.write(`data: ${JSON.stringify(chunk)}\n\n`);
> >             }
> >             res.write('data: [DONE]\n\n');
> >             res.end();
> >         }
> > ```
> 
> You can take reference from the code I shared in the thread.
> 
> ```js
> if (stream) {
>             res.setHeader('Content-Type', 'text/event-stream');
>             res.setHeader('Cache-Control', 'no-cache');
>             res.setHeader('Connection', 'keep-alive');
> 
>             const stream = await generateResponse(request); // should return response in OAI chat completion structure. return type: Array<response>
>             console.log('stream response', stream);  
>             for await (const chunk of stream) {
>                 res.write(`data: ${JSON.stringify(chunk)}\n\n`);
>             }
>             res.write('data: [DONE]\n\n');
>             res.end();
> }
> ```

thanks for the concept.. I am able to figure out how to send the output back to Deepgram from the custom LLM. But I have one Query, suppose the LLM link is incorrect Deepgram is not able to post the query and it is not showing or throwing any error. LLM server link is accessible from Deepgram that's why it is able to apply the settings. But unfortunately not able to post the text to custom LLM due to access issue. how to get the error out of Deepgram?
